src/interpreter.rs

```diff
@@ -1,148 +1,128 @@
 use std::collections::HashMap;
 
-fn throw (error: String) {
-    println!("{}", error);
-    std::process::exit(0);
-}
-
+#[derive(Debug)]
 pub enum Instruction {
     Increment,
     Decrement,
     MoveLeft,
     MoveRight,
-    StartLoop(u32), // u32 is the index of the end of the loop
-    EndLoop(u32),   // u32 is the index of the start of the loop
+    StartLoop(usize), // usize is the index of the end of the loop
+    EndLoop(usize),   // usize is the index of the start of the loop
     Output,
     Input,
     Halt,
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum BuildLoopMapError {
+    #[error("found loops with missing close instructions at indices {0:?}")]
+    StrayOpens(Vec<usize>),
+    #[error("found loop with missing open instructions at index {0}")]
+    StrayClose(usize),
+}
+
+// Creates mapping of loops and their endpoints for easy jumping around the code at loops endpoints and startpoints
+fn build_loop_map(code: &str) -> Result<HashMap<usize, usize>, BuildLoopMapError> {
+    use BuildLoopMapError::*;
+
+    let mut loops = HashMap::new();
+    let mut open_loops = Vec::new();
+
+    for (i, c) in code.chars().enumerate() {
+        match c {
+            '[' => open_loops.push(i),
+            ']' => {
+                let start = open_loops.pop().ok_or_else(|| StrayClose(i))?;
+
+                loops.insert(start, i);
+                loops.insert(i, start);
+            }
+            _ => {}
+        }
+    }
+
+    if !open_loops.is_empty() {
+        return Err(StrayOpens(open_loops));
+    }
+
+    Ok(loops)
+}

```

Comment:
moving this code to a standalone function allows writing test cases (whereas it was impossible before)
it's not exactly related to the Interpreter in any way